File: static/cpp/Merge_two_BSTs.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <memory>

#include "./BST_prototype_shared_ptr.h"

using std::cout;
using std::endl;
using std::make_shared;
using std::numeric_limits;
using std::shared_ptr;

shared_ptr<BSTNode<int>> BSTToDoublyListHelper(
    const shared_ptr<BSTNode<int>>&);

shared_ptr<BSTNode<int>> MergeTwoSortedLists(shared_ptr<BSTNode<int>>,
                                             shared_ptr<BSTNode<int>>);

void AppendNode(shared_ptr<BSTNode<int>>*, shared_ptr<BSTNode<int>>*);

// Build a BST from the (s + 1)-th to the e-th node in L.
shared_ptr<BSTNode<int>> BuildBSTFromSortedDoublyListHelper(
    shared_ptr<BSTNode<int>>* L, int s, int e) {
  shared_ptr<BSTNode<int>> curr = nullptr;
  if (s < e) {
    int m = s + ((e - s) / 2);
    auto temp = BuildBSTFromSortedDoublyListHelper(L, s, m);
    curr = *L;
    curr->left = temp;
    *L = (*L)->right;
    curr->right = BuildBSTFromSortedDoublyListHelper(L, m + 1, e);
  }
  return curr;
}

shared_ptr<BSTNode<int>> BuildBSTFromSortedDoublyList(
    shared_ptr<BSTNode<int>> L, int n) {
  return BuildBSTFromSortedDoublyListHelper(&L, 0, n);
}

shared_ptr<BSTNode<int>> BSTToDoublyList(const shared_ptr<BSTNode<int>>& n) {
  auto res = BSTToDoublyListHelper(n);
  res->left->right = nullptr;  // breaks the link from tail to head.
  res->left = nullptr;  // breaks the link from head to tail.
  return res;
}

// Transform a BST into a circular sorted doubly linked list in-place,
// return the head of the list.
shared_ptr<BSTNode<int>> BSTToDoublyListHelper(
    const shared_ptr<BSTNode<int>>& n) {
  // Empty subtree.
  if (!n) {
    return nullptr;
  }

  // Recursively build the list from left and right subtrees.
  auto l_head(BSTToDoublyListHelper(n->left)),
      r_head(BSTToDoublyListHelper(n->right));

  // Append n to the list from left subtree.
  shared_ptr<BSTNode<int>> l_tail = nullptr;
  if (l_head) {
    l_tail = l_head->left;
    l_tail->right = n;
    n->left = l_tail;
    l_tail = n;
  } else {
    l_head = l_tail = n;
  }

  // Append the list from right subtree to n.
  shared_ptr<BSTNode<int>> r_tail = nullptr;
  if (r_head) {
    r_tail = r_head->left;
    l_tail->right = r_head;
    r_head->left = l_tail;
  } else {
    r_tail = l_tail;
  }
  r_tail->right = l_head, l_head->left = r_tail;

  return l_head;
}

// Count the list length till end.
int CountLength(shared_ptr<BSTNode<int>> L) {
  int len = 0;
  while (L) {
    ++len, L = L->right;
  }
  return len;
}

// @include
shared_ptr<BSTNode<int>> MergeTwoBSTs(shared_ptr<BSTNode<int>> A,
                                      shared_ptr<BSTNode<int>> B) {
  A = BSTToDoublyList(A), B = BSTToDoublyList(B);
  int A_length = CountLength(A), B_length = CountLength(B);
  return BuildBSTFromSortedDoublyList(MergeTwoSortedLists(A, B),
                                      A_length + B_length);
}
// @exclude

// Merges two sorted doubly linked lists, returns the head of merged list.
shared_ptr<BSTNode<int>> MergeTwoSortedLists(shared_ptr<BSTNode<int>> A,
                                             shared_ptr<BSTNode<int>> B) {
  shared_ptr<BSTNode<int>> sorted_head(new BSTNode<int>);
  shared_ptr<BSTNode<int>> tail = sorted_head;

  while (A && B) {
    AppendNode(A->data < B->data ? &A : &B, &tail);
  }

  if (A) {
    // Appends the remaining of A.
    AppendNode(&A, &tail);
  } else if (B) {
    // Appends the remaining of B.
    AppendNode(&B, &tail);
  }
  return sorted_head->right;
}

void AppendNode(shared_ptr<BSTNode<int>>* node,
                shared_ptr<BSTNode<int>>* tail) {
  (*tail)->right = *node;
  *tail = *node;  // Resets tail to the last node.
  *node = (*node)->right;
}

template <typename T>
void PrintBSTInorder(shared_ptr<BSTNode<T>> n, const T& pre) {
  if (n) {
    PrintBSTInorder(n->left, pre);
    assert(pre <= n->data);
    cout << n->data << ' ';
    PrintBSTInorder(n->right, n->data);
  }
}

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4 6
  auto L = make_shared<BSTNode<int>>(BSTNode<int>{3});
  L->left = make_shared<BSTNode<int>>(BSTNode<int>{2});
  L->left->left = make_shared<BSTNode<int>>(BSTNode<int>{1});
  L->right = make_shared<BSTNode<int>>(BSTNode<int>{5});
  L->right->left = make_shared<BSTNode<int>>(BSTNode<int>{4});
  L->right->right = make_shared<BSTNode<int>>(BSTNode<int>{6});
  //     7
  //   2   8
  // 0
  auto R = make_shared<BSTNode<int>>(BSTNode<int>{7});
  R->left = make_shared<BSTNode<int>>(BSTNode<int>{2});
  R->left->left = make_shared<BSTNode<int>>(BSTNode<int>{0});
  R->right = make_shared<BSTNode<int>>(BSTNode<int>{8});
  shared_ptr<BSTNode<int>> root = MergeTwoBSTs(L, R);
  // should output 0 1 2 2 3 4 5 6 7 8
  PrintBSTInorder(root, numeric_limits<int>::min());
  return 0;
}

File: static/python/merge_two_bsts.py
from binary_tree_prototype import BinaryTreeNode


# Build a BST from the (s + 1)-th to the e-th node in L.
def build_bst_from_sorted_doubly_list_helper(L, s, e):
    curr = None
    if s < e:
        m = s + ((e - s) // 2)
        temp, L = build_bst_from_sorted_doubly_list_helper(L, s, m)
        curr = L
        curr.left = temp
        L = L.right
        curr.right, L = build_bst_from_sorted_doubly_list_helper(L, m + 1, e)
    return curr, L


def build_bst_from_sorted_doubly_list(L, n):
    return build_bst_from_sorted_doubly_list_helper(L, 0, n)[0]


# Transform a BST into a circular sorted doubly linked list in-place,
# return the head of the list.
def bst_to_doubly_list_helper(n):
    # Empty subtree.
    if not n:
        return None

    # Recursively build the list from left and right subtrees.
    l_head = bst_to_doubly_list_helper(n.left)
    r_head = bst_to_doubly_list_helper(n.right)

    # Append n to the list from left subtree.
    if l_head:
        l_tail = l_head.left
        l_tail.right = n
        n.left = l_tail
        l_tail = n
    else:
        l_head = l_tail = n

    # Append the list from right subtree to n.
    if r_head:
        r_tail = r_head.left
        l_tail.right = r_head
        r_head.left = l_tail
    else:
        r_tail = l_tail
    r_tail.right = l_head
    l_head.left = r_tail

    return l_head


def bst_to_doubly_list(n):
    res = bst_to_doubly_list_helper(n)
    res.left.right = None  # breaks the link from tail to head.
    res.left = None  # breaks the link from head to tail.
    return res


# Count the list length till end.
def count_length(L):
    length = 0
    while L:
        length += 1
        L = L.right
    return length


# @include
def merge_two_bsts(A, B):
    A, B = bst_to_doubly_list(A), bst_to_doubly_list(B)
    A_length, B_length = count_length(A), count_length(B)
    return build_bst_from_sorted_doubly_list(
        merge_two_sorted_lists(A, B), A_length + B_length)
# @exclude


# Merges two sorted doubly linked lists, returns the head of merged list.
def merge_two_sorted_lists(A, B):
    sorted_head = BinaryTreeNode()
    tail = sorted_head

    AB = [A, B]
    while all(AB):
        A_or_B = 0 if AB[0].data < AB[1].data else 1
        tail.right = AB[A_or_B]
        tail = tail.right  # Resets tail to the last node.
        AB[A_or_B] = tail.right

    if AB[0]:  # Appends the remaining of A.
        tail.right = AB[0]
    elif AB[1]:  # Appends the remaining of B.
        tail.right = AB[1]
    return sorted_head.right


def print_bst_inorder(n, pre):
    if n:
        print_bst_inorder(n.left, pre)
        assert pre <= n.data
        print(n.data, end=' ')
        print_bst_inorder(n.right, n.data)


def main():
    #      3
    #    2   5
    #  1    4 6
    L = BinaryTreeNode(3)
    L.left = BinaryTreeNode(2)
    L.left.left = BinaryTreeNode(1)
    L.right = BinaryTreeNode(5)
    L.right.left = BinaryTreeNode(4)
    L.right.right = BinaryTreeNode(6)
    #      7
    #    2   8
    #  0
    R = BinaryTreeNode(7)
    R.left = BinaryTreeNode(2)
    R.left.left = BinaryTreeNode(0)
    R.right = BinaryTreeNode(8)
    root = merge_two_bsts(L, R)
    # should output 0 1 2 2 3 4 5 6 7 8
    print_bst_inorder(root, float('-inf'))


if __name__ == '__main__':
    main()
